rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+code, the computation can then be run across multiple machines and / or devices.
+The computation that is running on each device is a [replica](#replica) of the
+user computation (assuming no [model parallelism](#model-parallelism) across
+devices).
+
+The Replicator API provides a number of primitives to facilitate communication
+across replicas. The most common use case is the aggregation of gradients for
+synchronous SGD training. Replicator natively supports this by automatically
+adding cross-replica gradient accumulation to any TensorFlow optimizer created
+within a replicator scope (see the
+[Simple Classification](#simple-classification) example below). For more
+advanced use cases (such as cross-replica batch normalisation), the API provides
+MPI-like primitives for sending and receiving tensors across replicas.
+
+The TF sub-graph generating input data is treated separately from the sub-graph
+performing the main computation (e.g. training a model). This decoupling
+provides flexibility in how input pipelines are connected to replicas. For
+example, replicas running on the same worker can be fed from the same input
+pipeline, which is often more efficient than having a separate input pipeline
+per replica.
+
+The API has been designed to work for all types of devices, including TPUs, so
+that code can be seamlessly moved between different device targets.
+
+### Concepts / Terminology
+
+#### Replica
+
+A replica is "one copy of the computation" that the user wants to replicate. You
+can think of it as "one copy of the model". In distribution strategy, this was
+formerly known as a "tower". We might also refer to this as "compute replica" or
+"model replica" in some places.
+
+When using model parallelism, a replica will span several physical devices (e.g.
+GPUs, TPU cores) and may also span multiple physical machines.
+
+#### Worker
+
+The physical machine(s) containing the physical devices (e.g. GPUs, TPUs) on
+which the replicated computation is executed. A worker may contain one or more
+replicas. Typically one worker will correspond to one host machine, but in the
+case of very large models with model parallelism, one worker may span multiple
+hosts.
+
+#### Sync vs Async replicas
+
+In synchronous training, all replicas are in sync with each other and gradients
+are aggregated across all the replicas. In fully asynchronous training, all
+replicas operate asynchronously and update the variables independently. We can
+also support a hybrid approach, where the replicas are divided into "sync
+groups" - within each sync group, each replica is in sync, but the sync groups
+are asynchronous among themselves. To support this, we later define
+"num_replicas" as the total number of replicas (sync or async) and
+"num_replicas_in_sync" as the number of replicas in one sync group.
+
+#### Input Pipeline
+
+An instance of the input pipeline (e.g. `Dataset`, `FIFOQueue`) obtained by
```

Comment:
Should Queues even be on the table? This would suggest they will be supported.

Reply:
I've removed it from this sentence as you're right, it might convey some special status for queues. I think we just meant that a queue might be used in an input function, but an example later shows this so I think we are good removing it from here.